### DISCUSSION TITLE
v86 with networking

### DISCUSSION THREAD

**Discussion by Nightlark-Darkmoon** [copy/v86#994] [General]
Hey! I am trying to implement a version of the 'two machines' example. I was wondering A. how to get it so that i could have two browser windows with two buildroot instances that can talk to each other? Then B. how is this networking handled, are they open to the internet? or is there something else on the backend that is handling the communciation.

**Comment by SuperMaxusa** [comment]
see #569

> A. how to get it so that i could have two browser windows with two buildroot instances that can talk to each other?

probably you need to use [Broadcast Channel API](https://developer.mozilla.org/en-US/docs/Web/API/Broadcast_Channel_API) to handle connection within browser tabs/windows that receive/send data to V86

> B. how is this networking handled, are they open to the internet?

In this example, when first instant sends data (`net0-send`), it's transfers to other instant then receive (`net0-receive`) that data on own network card. 
```
emulator1                                   emulator2
[   net0-send  ] ------------------> [ net0-receive ]
[ net0-receive ] <------------------ [   net0-send  ]
```
https://github.com/copy/v86/blob/360d5691ff2ef0ca1c1a16d683910338b7325f07/examples/two_instances.html#L68-L75

About public networking, this case is not suitable, atlease for this needed two "adapters" on each instant, and I don't have idea, how it can be done in v86. However, you can also connect two instants via same network relay: #783

**Reply by Nightlark-Darkmoon** [reply]
Ok So i think I understand what you are saying here. To clarify:

The net0-send and net0-recieve is contained entirely within the web browser and does not talk out to the open internet. And when there are two instances via network relay they live within their own vm /  network entirely contained within the web browser on which they are being served
